### ISSUE TITLE
Support configuration profiles

### ISSUE THREAD

**Issue by gastaldi** [eclipse/microprofile-config#418]
It would be nice if the spec allowed different profiles to be used when the application starts (eg. development, production, testing). The idea is that a different set of configuration properties would be available based on a flag (eg. `-Dprofile=production`).

In an hypothetical example, you could have a `META-INF/microprofile-config.properties` file with the default parameters and a `META-INF/microprofile-config-production.properties` with values used when the `production` value is specified which would override the default values.

**Comment by struberg** [comment]
Hi @gastaldi ! This is similar to what we had in Apache DeltaSpike ProjectStage, right?
http://deltaspike.apache.org/documentation/projectstage.html
It is rather easy to implement yourself with the freshly proposed `ConfigAccessor#addLookupSuffix`.
Ofc a bit more typing as with DeltaSpike. But as flexible.

**Comment by gastaldi** [comment]
@struberg yeah, that seems like a good solution. It would be nice if that were part of the spec instead of being an implementation detail.

**Comment by pilhuhn** [comment]
If implemented, please support some prefix for the property name like `-Dconfig.profile` and not only `-Dprofile`.

**Comment by Emily-Jiang** [comment]
I thougth this further. Maybe set the stage as part of property name is a better approach.
-Dstage=production
//stage=production
%production.name=hello
%live.name=world

**Comment by Emily-Jiang** [comment]
I started working on this and gave it more thought based on what @pilhuhn said.

- In order to avoid the name collison, I suggest use `mp.config.profile` or `mp.config.stage`. 

- Should we also support microprofile-config.properties and microprofile-config-testing.properties. I think many Spring users got used to this style and it is kind of easy fix. We only support microprofile-config.property but not other config sources. 

Thoughts?

**Comment by gastaldi** [comment]
> I started working on this and gave it more thought based on what @pilhuhn said.
> 
> 1. In order to avoid the name collison, I suggest use `mp.config.profile` or `mp.config.stage`.

+1, I like `mp.config.profile`

> 2. Should we also support microprofile-config.properties and microprofile-config-testing.properties. I think many Spring users got used to this style and it is kind of easy fix. We only support microprofile-config.property but not other config sources.

I don't have a preference here, I like different files as long as the additional properties file is used to override properties in the original config (a higher priority over the `microprofile-config.properties` ConfigSource). OTOH having part of the property name may be easier to maintain (that's what we have been using in Quarkus BTW).

**Comment by Emily-Jiang** [comment]
@gastaldi please take a look at the draft spec to see whether you have any more feedback.

**Comment by gastaldi** [comment]
Looks good. I've been thinking if the double `profile` in the feature name might not confuse users (eg. Micro*profile* Config *profile*)

**Comment by Emily-Jiang** [comment]
Good point @gastaldi ! `profile` is kind of overloaded. In 12 Factor APP, there was a mention of dev, staging, prod for the microservice lifecycle. Maybe we use `phase` to avoid clash with `staging`.
We have 3 names to choose from:

1. mp.config.profile
2. mp.config.stage
3. mp.config.phase
Can you all vote and choose one?

**Comment by gastaldi** [comment]
Let's add each option in separate comments for voting

**Comment by gastaldi** [comment]
`mp.config.profile`

**Comment by gastaldi** [comment]
`mp.config.stage`

**Comment by gastaldi** [comment]
`mp.config.stage`

**Comment by Emily-Jiang** [comment]
ok. The winner is `mp.config.profile`, which is my vote as well:o.

**Comment by Emily-Jiang** [comment]
I would like to add another vote for config property name convention. Let's vote with the following options.

**Comment by Emily-Jiang** [comment]
`%profile%property.name`

 (e.g. `%staging%vehicle.name`)

**Comment by Emily-Jiang** [comment]
%profile.propertyName

(e.g. %staging.vehicleName)

**Comment by dmlloyd** [comment]
As a follow-up note, I don't think Quarkus is likely to support this spec feature if we don't go with the standard `%profile.propertyName` syntax.  So I guess in this case we would stop complying with MP Config and possibly walk away from it for good.

**Comment by Emily-Jiang** [comment]
@dmlloyd there is a delema with the implementation first. If there is another implementation said the same, MP will ground to hold or let implementations to walk away. Surely, in SmallRye, you could add a shim layer to translate or support both. 
Having that said, since this case is pretty minor, there is not much distinction between the two. I am ok to adopt `%profile.propertyName`

**Comment by dmlloyd** [comment]
In this case there are no other implementations.  But yes, eventually we are going to have a serious problem.

**Comment by Emily-Jiang** [comment]
My point is that what defined in MP should not be limited by implementations. It has great flexibility to explore different ways to solve the problems.

**Comment by smillidge** [comment]
I am confused by this discussion. In the first vote are you talking about the name of the System property? Also what happened to the discussion on separate files? I would prefer separate files rather than a single file with multiple scoped property names. Is separate files in or out?

**Comment by radcortez** [comment]
We can also support separate files.

The motivation to come up with multiple scoped property names is to be able to use profile for any ConfigSource. With a separate file approach, you cannot use profile configuration for other sources like databases, vault, etc.

**Comment by smillidge** [comment]
but as they are external sources why would you store profiles in them they would just be configured differently in different environments? Saying that supporting property names brings added flexibility but could lead to some confusion with string clashes if a developer has inadvertently included a profile string in their property name.  
What are the resolution rules? 
How do you determine which portion of the name is the profile and which piece is just how the developer named the property?

**Comment by radcortez** [comment]
The idea is to prefix a property with `%` followed by the profile name and a dot. So you can have:

```
my.prop=1234
%dev.my.prop=5678
```

And you use `mp.config.profile=dev` to activate it. So you resolve a profile my prefixing `%dev` in the lookup of the config. I believe that is very unlikely for developers to have properties starting with `%`.  Yes, it may happen.

Do you think this shouldn't be supported at all?

**Comment by OndroMih** [comment]
I think we should evaluate how profiles are supported by existing configuration frameworks, like Spring Configuration or DeltaSpike config. We shouldn't invent anything that isn't used anywhere by widely adopted frameworks. And however I respect Quarkus, it's been around just for 1 year, so it's not a good enough reference to standardize functionality in MP Config.

**Comment by radcortez** [comment]
This is an addition.

No one said that profile based files couldn't be supported. I see no issue in supporting both.

**Comment by radcortez** [comment]
Also, let me add that one of my previous employers had limited resources that translated in same sources (maybe vault or database) serving different environments (dev, stg, qa, etc. prod separate of course :)).

I've found this really useful in that situation. I agree that we cannot standardise based on one person opinion, but on the other hand we cannot just assume that everyone out there has perfect environments either :)

**Comment by OndroMih** [comment]
I understand, @radcortez, that supporting profiles in property names is an addition to supporting profiles for properties files. However, we should carefully evaluate if it's standard enough to add it to MP Config and how to add it so that it aligns best with the existing widely used config libraries.

As an example, Spring only supports profiles for whole files, not for each separate property: https://docs.spring.io/spring-boot/docs/2.2.6.RELEASE/reference/html/spring-boot-features.html#boot-features-external-config. 

On the other side, DeltaSpike Config doesn't support separate files for each profile. It supports properties with a project stage suffix and allows retrieving properties based according to the project stage or ignoring project stage. Injected properties resolve projecct stage by default but the annotation has a flag to disable this.

Apache Commons config doesn't support any of this directly, it only supports multi-tenant configuration, which is somewhat similar to supporting profiles for whole files: http://commons.apache.org/proper/commons-configuration/userguide/howto_multitenant.html#Multi-tenant_Configurations

cfg4j library supports profiles for whole files. It calls it "environments" and files for different environments are read from different directory. Each config source can support the supplied value for environment as they want or just ignore it. Docs for file config source: https://www.javadoc.io/doc/org.cfg4j/cfg4j-core/4.4.0/index.html

So, in summary, for the 4 most popular config libraries I found, 3 only support profiles for whole files and one (DeltaSpike) supports profiles for individual properties. This doesn't look like worth standardizing in MP Config.

**Comment by OndroMih** [comment]
@radcortez, I think that the case with your former employer is rather an edge case and still can be supported by a custom config source. Not to mention that to support config in DB a custom config source is needed anyway.

I propose that we standardize the config property that specifies the profile name and define how this property is interpreted by the standard properties config source defined by MP (e.g. searching for files with the profile prefix in their name). Other custom config sources can use the value of the standard property to implement support for profiles as they want.

I don't see a value of supporting profiles in env vars and external system properties. And then there are no other standard config sources defined in MP Config.

**Comment by radcortez** [comment]
Well, one question that we need to ask ourselves is that we just want to standardise things and copy from other API's or implementations? Or do we want to innovate as well?

**Comment by radcortez** [comment]
> @radcortez, I think that the case with your former employer is rather an edge case and still can be supported by a custom config source.

Yes. You can support a lot of things with custom config sources and while that works, I believe this is not ideal. The reason why you do this, is because there is no other extension point to use, so you need to become creative and use the hammer for every thing you are trying to do.

**Comment by OndroMih** [comment]
The question wehther we just want to standardize things from existing implementations or add innovate is already being discussed here: https://groups.google.com/d/msg/microprofile/G5WZvF0wZuY/BAGsTRCABAAJ. We should resolve that first for MP globally before we add new innovative features that are not proven by widely used projects.

**Comment by radcortez** [comment]
I'm aware of that discussion and even if there is a general guideline to follow globally, I think that things can neither be black or white. There is always some grey in between.

In this particular case, we are not talking about a big change or something that will break everyone. If we limit ourselves to just copy stuff without considering other things, either this feature or something else, I believe we are quickly going to loose interest from users.

**Comment by smillidge** [comment]
The purpose of MP Config is to externalise configuration from the application. As @OndroMih says we only have 3 config sources defined; system properties, environment variables and embedded  properties files. The first two are external to the application and therefore can be externally configured for different environments. The classpath properties file(s) is really the only one source that couldn't support different environments. However to support it a developer would still be adding in environment specific property files into the application which IMHO is a bad practice as they can't really know up front all the different environments an application can be "deployed" to.

I'm not sure how adding some replaceable identifier to a property name really buys much.

**Comment by dmlloyd** [comment]
Using separate "files" requires active participation from every ConfigSource implementation. Whereas the property name approach already works with most, if not all, existing implementations.  So consider what you are proposing.  Must every existing implementation now rewrite every config source?

**Comment by smillidge** [comment]
The files was suggested at the top by the requestor and solves the specific issue for the config source that uses properties files embedded on the classpath specified in MP Config. Other "files" config sources are not specified as part of MP Config and I don't see why "files" impacts say a DB config source or the env or system property config sources?

**Comment by dmlloyd** [comment]
Because each source would need specific knowledge of the profile in use to know what "file" to open. Today no config source is required to have that knowledge.

**Comment by dmlloyd** [comment]
Otherwise we don't have uniform support - it becomes as-hoc depending on th config source.

**Comment by radcortez** [comment]
> The purpose of MP Config is to externalise configuration from the application. As @OndroMih says we only have 3 config sources defined; system properties, environment variables and embedded properties files. The first two are external to the application and therefore can be externally configured for different environments. The classpath properties file(s) is really the only one source that couldn't support different environments. 

Hum... if we want to externalise the embedded classpath properties file, you don't need profiles for that, right? Are you proposing to add a way to externalize it and drop any profile support?

**Comment by smillidge** [comment]
My point is, only the config source that reads META-INF/microprofile-config.properties requires this profile support (if it requires it at all) every other config source is obtaining externally defined configuration and therefore does not require profile support to be specified.

**Comment by smillidge** [comment]
To make it more concrete if I was have a Kubernetes secrets config source and I wanted to switch environments and thereby the property values injected in my application then I can change the secrets values in Kubernetes there is no need to mess with -Dmp.config.profile and %foo.bar property names. That is the point of MP Config or am I missing something?

**Comment by radcortez** [comment]
If the goal is to only externalise the file, then this is not the right approach.

Profiles have other use cases, for instance  multi-tenancy applications. With profiles, I can create and load Configs for different tenants, while keeping the API rather simple.

**Comment by dmlloyd** [comment]
Maybe we have different understandings of the use case for profiles.  In Quarkus profiles apply to every configuration source type already.  Profile-specific names take precedence over the non-profile name in the same config source or ordinal.  This allows users to define a potentially large and complex configuration - including across custom configuration sources - but only override a small subset of properties on a per-profile basis (e.g. development, testing, production).  The configuration source does not need any knowledge of profiles, which makes sense because it's a separate concern.

If the multiple files approach is used, then the configuration is completely separate and must be duplicated to the separate file.  If both the profile file *and* the non-profile file are read, in addition to putting more burden of complexity on the config source implementation, it also becomes impossible for the user to place profile-specific configuration alongside the profile-agnostic configuration it relates to.  And of course as I said every config source needs explicit support for this.

The brief history of the Quarkus approach: essentially we saw what Play was doing (for the reasons I gave above), so we went on with a similar approach.

But perhaps more importantly, if every implementation has a conflicting approach for profiles, then perhaps this proposal should be withdrawn altogether and we all go our separate ways with it.

**Comment by smillidge** [comment]
I think we are talking at cross purposes. 

My belief is that MP Config externalises configuration so where the configuration values comes from is no longer a concern of the application developer. Adding in profile support via system properties and %foo replaceable property names is just adding more application developer concerns and therefore is internalising configuration again. 

Setting environment/tenant specific properties should be up to the "deployer" of the application whether the deployer is configuring kubernetes, docker or an application server.

**Comment by Emily-Jiang** [comment]
@OndroMih The profile concept is not new. It exists in Spring and has a different form in DeltaSpike Config in the form of staging, which is property based.

If the profile property just override the property in the same cnfigsource, what if the config property does not exist? e.g.

Configsource 1: ordinal:310
%dev.vehicle.name=car
vehicle.name=bike

configsource 2: ordinal: 320
%dev.vehicle.name=van

When I do the following injection:
`@Inject @ConfigProperty(name="vehicle.name") String vehicleName;
@dmlloyd I expect the value for `vehicleName` is `van` though. Do you agree? If not, which value do you expect to see?

**Comment by radcortez** [comment]
I see environment and tenants as two different things.

An application runs in single environment at a time (dev, qa, prod, etc), but may have multiple tenants using the application (companyA, companyB, etc)., with their own separate data and users.

Consider a `page.size` attribute set per organization (tenant). I think it is useful to pass an additional discriminator (profile) so I can retrieve the specific configuration for that organization. I may have `microprofile-config-companyA.properties`,  `microprofile-config-companyB.properties`, etc. and create different instances of Configs in my runtime to load the specific configuration.

Do you see a use case here?

**Comment by smillidge** [comment]
But MicroProfile Config solves the use case now. If you want to override properties in META-INF/microprofile-config.properties as you move between environments you can set them as system properties or environment variables.  

Obviously outside the specification runtimes can provide additional ConfigSources that offer more flexibility.

What is suggested at the moment is not much more really than an opaque hint to a config source that it should rewrite property names on the fly. It adds complexity to the specification for little utility.

**Comment by smillidge** [comment]
@radcortez I don't think any MP specification should concern itself with multi-tenancy. Also I don't think the current suggestion would satisfy your use case. 

Also @dmlloyd I don't think MP Config should have `the configuration configures the frameworks` as a goal either. That path will lead to a lot of complexity in the API of little utility to the average  application developer.

**Comment by dmlloyd** [comment]
> Also @dmlloyd I don't think MP Config should have `the configuration configures the frameworks` as a goal either. That path will lead to a lot of complexity in the API of little utility to the average application developer.

That is irrelevant, because it is already the reality that this is being done.

**Comment by radcortez** [comment]
> @radcortez I don't think any MP specification should concern itself with multi-tenancy. Also I don't think the current suggestion would satisfy your use case.
> 

This is not about the `%profile` discussion anymore. Is about profiles as an overall mechanism. That use case is just an example and a profile feature, (including property aware profiles or not) is able to implement the use case.

I think we need to define what a "profile" means, because it seems all of us have a different view of what a profile is.

**Comment by smillidge** [comment]
> 
> 
> > Also @dmlloyd I don't think MP Config should have `the configuration configures the frameworks` as a goal either. That path will lead to a lot of complexity in the API of little utility to the average application developer.
> 
> That is irrelevant, because it is already reality that this is being done.

I don't want to drive this off topic so this would a separate conversation but MP Config would need a lot of additional features to support framework configuration that are not relevant to end user developers. Payara would have loads of additional requirements. However MP Config should not be distracted by framework configuration requirements.

**Comment by dmlloyd** [comment]
> > > Also @dmlloyd I don't think MP Config should have `the configuration configures the frameworks` as a goal either. That path will lead to a lot of complexity in the API of little utility to the average application developer.
> > 
> > 
> > That is irrelevant, because it is already reality that this is being done.
> 
> I don't want to drive this off topic so this would be a separate conversation but MP Config would need a lot of additional features to support framework configuration that are not relevant to end user developers. Payara would have loads of additional requirements. However MP Config should not be distracted by framework configuration requirements.

Why not?

**Comment by Emily-Jiang** [comment]
Further to our chat today with @radcortez and @jbee
I think the ordinal does come to play with config profile. See the following example:

If the profile property just override the property in the same cnfigsource, what if the config property does not exist? e.g.

Configsource 1: ordinal:310
```
%dev.vehicle.name=car
vehicle.name=bike
```

configsource 2: ordinal: 320
`%dev.vehicle.name=van` 



-- I assume without the `vehicle.name` property, this property will be still be read.

1. When I set `mp.config.profile=dev` and do the following injection:

`@Inject @ConfigProperty(name="vehicle.name") String vehicleName;
@dmlloyd I expect the value for `vehicleName` is `van` though. Do you agree? If not, which value do you expect to see? If yes, the higher ordinal wins.

2. When I set `mp.config.profile=live` and do the following injection:

`@Inject @ConfigProperty(name="vehicle.name") String vehicleName;
@dmlloyd I expect the value for `vehicleName` is `bike` though. Do you agree? If not, which value do you expect to see?

3. If I add another config source
configsource 3: ordinal: 330
`vehicle.name=train` 

`@Inject @ConfigProperty(name="vehicle.name") String vehicleName;
@dmlloyd I expect the value for `vehicleName` is `train` though. Do you agree? If not, which value do you expect to see?
In this way, the config ordinal still comes to play.

**Comment by dmlloyd** [comment]
Correct, highest ordinal wins; if ordinals are equal then profile-specific name wins.

**Comment by OndroMih** [comment]
@radcortez, multitenancy and deployment profiles are very similar in fact. They both require to define config values for different deployments/tenants and to be able to specify which one of them is used.

There are also differences because multitenance has more requirements:
* multitenancy requires that the application can switch between tenants based on runtime conditions while deployment profiles are set once for a deployment
* the number of tenants can be potentially dynamic, it can grow or shrink during the lifetime of the application

**Deployment profiles** can be easily implemented by a separate file for each profile **inside the app** and set the active profile during deployment, e.g. by a system property, env var or an implementation-specific deployment option. This is what most config solutions support. Usually it's not needed to switch between different profiles in the same environment. The environment already provides profile-specific config and thus support for profiles are not needed in external config.

**Multitenancy**, however, requires that it's possible to switch between profiles during runtime and thus each config source, including external sources, should support profiles. One solution to do this is to support profile prefix in property names as suggested above. 

Do we want to support deployment profiles or multitenancy? 

This issue was raised to support deployment profiles and we should focus on that to keep things simple. If somebody is proposing to support multitenancy, I suggest they raise a separate issue.

**Comment by OndroMih** [comment]
@Emily-Jiang, I know that profiles aren't new and  above I presented results of my research of how existing solutions support it. Please look it up in my previous comments. DeltaSpike is the only one that supports profiles on property level. 

Before adding a similar thing to MP Config, I would consult DeltaSpike team first to tell us what they think about their decision to support it and whether they would do something differently. @struberg, would you like to comment on that?

**Comment by gastaldi** [comment]
@OndroMih IMHO using prefixes in the property names or having separate files per profile are mere implementation details and have no relation if are switched during run-time.

After using this feature in Quarkus for quite some time now, I prefer having a prefix in the property name instead of separate files, mostly because it's easier to maintain - and I haven't heard of anybody complaining so far about this strategy :wink:

**Comment by dmlloyd** [comment]
> @radcortez, multitenancy and deployment profiles are very similar in fact. They both require to define config values for different deployments/tenants and to be able to specify which one of them is used.

The original issue defined profiles as "a flag" which allows "a different of configuration properties [to be available]", and the subsequent discussion has been based on that definition.  I don't know where your definitions of these terms ("profile" and "multitenancy") come from but I can say for sure that they are not absolute/global definitions.

It is IMO disingenuous and unhelpful to retroactively redefine what is being asked for, and then suggest that the scope of what is being requested is smaller than what the original issue asked for as a result.

**Comment by Emily-Jiang** [comment]
@OndroMih @struberg commented [this](https://github.com/eclipse/microprofile-config/issues/418#issuecomment-472323597) earlier and mentioned deltaspike. I think we are in agreement that config profile does not cause confusion and only wins over the normal config if it's has the same ordinal as the normal config. You can continue override your property as before. By the way, even adding this in the spec, end users don't even need to use it if they don't need to. It is quite easy to implement.
In addition to this, I plan to also add another PR to do the spring config style of config profile by introducing microprofile-config-${config.profile}.properties, as Spring developers got used to that style but it is quite limited and not as flexible as the property level profile support. However, sometimes it is quite to copy and override some values and rename the file.

**Comment by smillidge** [comment]
As a point of order using terms as `disingenuous` is an attack on the individual and their motives. It does not help promote collaborative working or a friendly atmosphere.

**Comment by OndroMih** [comment]
Hi @Emily-Jiang, I like the idea of raising 2 separate PRs for both options. Then we can discuss them separately.

I saw the comment from @struberg earlier, but the comment suggested that it can already be done with another feature and he didn't confirm whether it's worth standardizing:

> It is rather easy to implement yourself with the freshly proposed ConfigAccessor#addLookupSuffix.
> Ofc a bit more typing as with DeltaSpike. But as flexible.

But I also acknowledge that @gastaldi likes the solution with profiles for individual properties although he initially suggested a simpler solution with multiple files.

I won't comment @dmlloyd anymore, at least until he apologizes to me privately. Personal opinions like those shouldn't be posted here.

**Comment by radcortez** [comment]
I didn't want to start a discussion about multi-tenancy or anything else. I only wanted to show a use case where a profile at the property level could be used. There are other use cases and it's fine to disagree/agree on any of them.

Unfortunately, we are just a few individuals that represent a large community of users, which usually don't voice their opinions here. We can never be sure if what we propose or add is something that users want, and yes, looking into the state of the art help and what others are doing may give us a good indication.

On the other hand, and even if I don't want to start the discussion about innovation/stability since there is a thread for that, I believe we should go the extra mile if we can. We have very smart people here and I think it is a waste to use these people just to copy things that others did.

Moving forward, if we want to have 2 PR's to discuss things separately, this is fine. On the other hand, PR's will consist of the Spec and TCK, I don't see any API work required. Both PR's most likely are going to overlap or touch the same portions which may cause some confusion. What we really need to agree on is the spec requirement. There is already a PR here: https://github.com/eclipse/microprofile-config/pull/554 by @Emily-Jiang. The PR has 40 lines of the spec to cover everything. Do we really need to split it?

Also, we have a spec lead, @Emily-Jiang. I'm not completely about the role responsibility in the context of MP projects, but it seems to me that if there is no consensus on how this should be handled (1 PR vs 2 PR atm), we should let @Emily-Jiang decide on how to handle it.

**Comment by OndroMih** [comment]
Hi @radcortez , first, I believe that @Emily-Jiang is perfectly capable of leading MP Config and I respect her opinion very much. But I can't agree that he has any authority from a role of a spec lead. MicroProfile is very different from JCP and there are no spec leads like in JCP. Everybody has equal ownership of the spec and we need to remember that all spec implementers have to implement what we agree on here. Therefore we have to be careful about adding too much innovation too rapidly, especially if there's not a broad consensus to support such innovation.

That said, I'll continue the discussion on PR #554 so that we can move somewhere. We've discussed this long enough. Now it's important to have a real and good proposals to vote on.

**Comment by radcortez** [comment]
Sure, but there are going to be occasions that we are not able to reach a consensus and we cannot be blocked either. I didn't propose for @Emily-Jiang to decide on the entire feature, just to decide to either do it in a single PR or separate in multiple PRs. She is PR author anyway, so I believe she has the authority to do it.

**Comment by Emily-Jiang** [comment]
@OndroMih @radcortez thanks for your comments! Both of you are correct. I am not going to split the current PR #554 as it covers the whole piece of config property level of profile support. What I am going to create a PR is for file based config profile. They are complimentary not contradicte to each other to cover the following scenario.

A jar containing `microprofile-config.properties`, `microprofile-config-dev.properties`, `microprofile-config-prod.properties` under META-INF dir.

If `mp.config.profile=dev`, microprofile-config-dev.properties while the other two files are ignored.

**Comment by arjantijms** [comment]
Sidestepping the current discussion a little, it's perhaps interesting to see that for the pre-cursor of the pre-cursor of MP Config (the pre-cursor of Java EE Config) something with the same effect was proposed:

> [...]
> A possible solution for this could be the introduction of descriptor fragments that are included from the main descriptor based on a placeholder. E.g. in web.xml:
> 
> ```xml
> <web-app> 
>     ... 
>     <fragment>WEB-INF/conf/${mycompany.staging}/web-fragment.xml</fragment>
> </web-app>
> ```
> 
> Inside the WEB-INF/conf directory, multiple folders could be created, each corresponding to a stage, e.g.
> 
> ```
> WEB-INF
>     conf
>         dev
>             web-fragment.xml
>         qa
>             web-fragment.xml
>         live
>             web-fragment.xml
> ```
> 
> Starting up the application with `-Dmycompany.staging=dev` would then cause `WEB-INF/conf/dev/web-fragment.xml` to be processed.

http://web.archive.org/web/20141018170035/https://java.net/jira/browse/JAVAEE_SPEC-19

**Comment by smillidge** [comment]
@Emily-Jiang RE: `microprofile-config-${config.profile}.properties` being a copy of all the values in the standard microprofile-config.properties you could have both profile specific file(s) searched first then the standard file(s).

**Comment by Emily-Jiang** [comment]
> @Emily-Jiang RE: microprofile-config-${config.profile}.properties being a copy of all the values in the standard microprofile-config.properties you could have profile specific file(s) searched first then the standard file(s).

correct @smillidge ! I saw some end users wanted to directly copy the whole file and only change a couple of values and rename the file. It is not as powerful as the config profile on property level but it will cover some use cases, especially for Spring developers.

**Comment by Emily-Jiang** [comment]
I'll propose a file based config profile, similar to what Spring config supports. I propose to support this for `microprofile-config.properties`. You can have `microprofile-config.properties,` `microprofile-config-production.properties` or `microprofile-config-testing.properties`. If `mp.config.profile` is `production`, the file `microprofile-config-production.properties` is used while the other ones are ignored. The file `microprofile-config.properties` is only a fallback option. Thoughts?

**Comment by radcortez** [comment]
Do we want to assign actual names with associated behaviour? (prod being active if exists).

As far as I know, in Spring you still need to activate the files via a profile:
https://docs.spring.io/spring-boot/docs/current/reference/html/spring-boot-features.html#boot-features-external-config-profile-specific-properties

**Comment by Emily-Jiang** [comment]
> Do we want to assign actual names with associated behaviour? (prod being active if exists).

@radcortez can you explain a bit more on this?

> As far as I know, in Spring you still need to activate the files via a profile:
> https://docs.spring.io/spring-boot/docs/current/reference/html/spring-boot-features.html#boot-features-external-config-profile-specific-properties

What I suggested above retains the same behaviour as Spring. 

At the moment, under META-INF/ folder, only `microprofile-config.properties` is loaded at the moment. If you add other files, you have to add that configsource via service loader. If you place microprofile-config-prod.properties under META-INF, it will be ignored. With the proposed change, if `mp.config.profile` is set to `prod`, the file `microprofile-config-prod.properties` will be loaded while `microprofile-config.properties` is the backup.

**Comment by radcortez** [comment]
Ah fine, I think I misunderstood the previous comment.

It does seem that when you load a profile, the default file is not used at all.
